otter/test/convergence/test_composition.py

```diff
@@ -25,25 +70,33 @@ def setUp(self):
         Sample server json
         """
         self.servers = [
-            {'id': 'a', 'state': 'ACTIVE', 'created': now(),
-             'addresses': {'private': [{'addr': 'ip1', 'version': 4}]}},
-            {'id': 'b', 'state': 'ACTIVE', 'created': now(),
-             'addresses': {'private': [{'addr': 'ip2', 'version': 4}]}}
+            NovaServer(id='a', state=ServerState.ACTIVE, created=0, servicenet_address='ip1'),
+            NovaServer(id='b', state=ServerState.ACTIVE, created=0, servicenet_address='ip2'),
         ]
 
+    def _get_get_all_convergence_data(self, servers, group_id, reqfunc):
```

Comment:
Well, here are the options I know of, and which are really a generalization of the whole problem of testing, dependency injection, and abstraction.

- don't parameterize it, and just patch/mock it in the tests.
- parameterize it, which is really only marginally more reliable than patch/mocking...
- don't parameterize it at all, all the way down to get_scaling_group_servers et al, and have the tests check all the Effects of HTTPRequests that they'll get back. this is pretty undesirable since it means effectively duplicating a lot of the tests for those lower level functions.
- abstract the whole thing with a new Intent that describes fetching scaling group servers, etc. This is a big step that's kind of tedious and one I'm wary to do too much, as it leads to a massive proliferation of new abstractions. Maybe I'm too wary, and it's a great idea, but I'm actually worried that I've been seeing too many nails with this hammer lately (ServiceRequest, Authenticate, etc).

I am quite open to the idea that I've been getting stuck in a particular pattern of thinking, so if you have any recommendations, I'm happy to hear them :)